### ISSUE TITLE
Service definition(cognito-idp): Should `UserPoolAddOnsType.AdvancedSecurityMode` be optional?

### ISSUE THREAD

**Issue by endgame** [boto/botocore#2514]
I am a maintainer of Amazonka, an unofficial Haskell SDK for AWS. We use botocore definitions to generate the bulk of the Haskell code. In https://github.com/brendanhay/amazonka/issues/473 , a user reported that calling `CreateUserPool` with no `UserPoolAddOns` field set (as permitted by [API spec](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_CreateUserPool.html#CognitoUserPools-CreateUserPool-request-UserPoolAddOns)) resulted in a deserialisation failure. This was because the response appeared to contain `"UserPoolAddOns": {}`.

We have worked around this by making the `AdvancedSecurityMode` setting optional, so that `{}` can be parsed as a `UserPoolAddOnsType`, but it is not clear what the correct upstream fix is. All the remedies i can think of either duplicate definitions, admit malformed requests, or risk breaking clients who rely on the response from `CreateUserPool` to always include the `UserPoolAddOns` key in its `UserPool` structure.

**Comment by endgame** [comment]
https://github.com/boto/botocore/blob/77359646878e261eaa9debad888fc8f3b20832ef/botocore/data/cognito-idp/2016-04-18/service-2.json#L7758-L7760

**Comment by stobrien89** [comment]
Hi @endgame,

Thanks for your patience!  From everything I can see, it appears `AdvancedSecurityMode` should be required for `UserPoolAddOnsType`.  Do you know if the user was including the option in their `CreateUserPool` request?  The behavior I'm seeing is that, upon user pool creation, if `UserPoolAddOns` is not included in the request, the field does not appear in the response.  When providing `UserPoolAddOns` in the request, we have client-side validation in place to ensure a user is specifying `AdvancedSecurityMode` and cognito has validation in place to ensure the values provided for `AdvancedSecurityMode` are one of the following: `[ENFORCED, OFF, AUDIT]`.  I'll bring this up with the team tomorrow to get their thoughts as well.

**Comment by endgame** [comment]
Thanks for getting back to me @stobrien89 . Unfortunately this has come from digging through years-old bugs so I can't go back to my reporter and ask for more info. `AdvancedSecurityMode` is required for `UserPoolAddOnsType`, but `UserPoolAddOnsType` itself appears to be optional in the `CreateUserPool` request.

But if it's omitted in the request, it appears that the response generates `UserPoolAddOns: {}`, which contradicts the service definition. If this is true, there are several remedies I can see:

1. Stop returning `UserPoolAddOns: {}` if it wasn't provided in the request. May break downstream clients.
2. Use a different shape in the request vs. response type. Causes proliferation of shape types for minimal benefit.
3. In boto, declare `AdvancedSecurityMode` to be optional in `UserPoolAddOnsType`. Allows malformed requests.
4. In amazonka, declare `AdvancedSecurityMode` to be optional in `UserPoolAddOnsType`. We have infrastructure to deal with this sort of thing, and while it's regrettable that we can't upstream every error that we find in service definitions, it might cause the least breakage. Generally, amazonka users can't rely on the types in generated code; they have to read AWS docs anyway.

I think that in an ideal world, if I'm right about the cause, (1) is probably the "most right" fix, but it risks breaking an unknowable amount of downstream code. (4) may be the most pragmatic fix, unfortunately; it'll be interesting to see what your team says.

**Comment by stobrien89** [comment]
Hi again @endgame,

I reviewed this with the team yesterday and they confirmed that we don't actually validate responses from a given service (except in the case of some customizations), so the required fields are, from a functional standpoint in the Python SDK and the AWS CLI, largely to validate input.  

I'm not sure how Cognito would handle an empty object for `UserPoolAddOns` sent in the request body, so I thought that may have been the case for the user you mentioned.  Like I mentioned before, I'm not seeing `UserPoolAddOns` in the raw response when I exclude it from the request.

**Comment by endgame** [comment]
I created a user pool with a body of `{"PoolName":"test-pool"}` in the request and get no `UserPoolAddOns` in the response. If I try `{"PoolName": "test-pool", "UserPoolAddOns": {}}` in the request, it is correctly rejected by AWS.

Based off our bug report, the request was reported to have no `UserPoolAddOns` and the response `UserPoolAddOns: {}`, but I cannot replicate this in either us-east-1 or ap-southeast-2. If Cognito was ever returning a response that disagreed with its own service definition, it has long since been fixed. Thanks very much for your time.

Closing - not a bug?
